### PR TITLE
Fix `impBoxPatternMatch`

### PR DESCRIPTION
**Pull request by SingleAccretion**
For the `BOX;BR[COND]` case, the issue was that the code assumed that we only need to hold onto the original address tree in case a null check is needed, which is not the case -- the address could be both non-null and itself side-effecting (e. g. `ADDR(INDEX)`).

[Diffs](https://dev.azure.com/dnceng/public/_build/results?buildId=1765495&view=ms.vss-build-web.run-extensions-tab).

### CODE REVIEW

File: src/coreclr/jit/importer.cpp [hunk 1]
```diff
@@ -6952,50 +6952,22 @@ int Compiler::impBoxPatternMatch(CORINFO_RESOLVED_TOKEN* pResolvedToken,
...
-                        }
-                    }
-                    else
-                    {
-                        canOptimize = false;
-                    }
-                }
-
-                if (canOptimize)
+                CorInfoHelpFunc boxHelper = info.compCompHnd->getBoxHelper(pResolvedToken->hClass);
```
Comment (jakobbotsch) [inline src/coreclr/jit/importer.cpp:6955]: You should avoid asking for this helper at all for `opts == BoxPatterns::IsByRefLike`.

**Comment by SingleAccretion** [comment]
@dotnet/jit-contrib I would like to request a re-run of the SPMI diffs for this change. It [failed](https://dev.azure.com/dnceng/public/_build/results?buildId=1733279&view=logs&jobId=011e1ec8-6569-5e69-4f06-baf193d1351e&j=011e1ec8-6569-5e69-4f06-baf193d1351e&t=bf6cf4cf-6432-59cf-d384-6b3bcf32ede2) to download the Jits from the rolling build it seems:
```
Downloading:

  https://clrjit2.blob.core.windows.net/jitrollingbuild/builds/4198a8af7caa3f9f28db68a90ef4b318f7edb50c/windows/x64/Checked/clrjit.dll

  https://clrjit2.blob.core.windows.net/jitrollingbuild/builds/4198a8af7caa3f9f28db68a90ef4b318f7edb50c/windows/x64/Checked/clrjit_unix_x64_x64.dll

  https://clrjit2.blob.core.windows.net/jitrollingbuild/builds/4198a8af7caa3f9f28db68a90ef4b318f7edb50c/windows/x64/Checked/clrjit_win_x64_x64.dll

  https://clrjit2.blob.core.windows.net/jitrollingbuild/builds/4198a8af7caa3f9f28db68a90ef4b318f7edb50c/windows/x64/Checked/clrjit_win_x86_x64.dll

Download: https://clrjit2.blob.core.windows.net/jitrollingbuild/builds/4198a8af7caa3f9f28db68a90ef4b318f7edb50c/windows/x64/Checked/clrjit.dll -> D:\a\_work\1\s\payload\base\clrjit.dll

Download: https://clrjit2.blob.core.windows.net/jitrollingbuild/builds/4198a8af7caa3f9f28db68a90ef4b318f7edb50c/windows/x64/Checked/clrjit_universal_arm64_x64.dll -> D:\a\_work\1\s\payload\base\clrjit_universal_arm64_x64.dll

Download: https://clrjit2.blob.core.windows.net/jitrollingbuild/builds/4198a8af7caa3f9f28db68a90ef4b318f7edb50c/windows/x64/Checked/clrjit_universal_arm_x64.dll -> D:\a\_work\1\s\payload\base\clrjit_universal_arm_x64.dll

Download: https://clrjit2.blob.core.windows.net/jitrollingbuild/builds/4198a8af7caa3f9f28db68a90ef4b318f7edb50c/windows/x64/Checked/clrjit_unix_x64_x64.dll -> D:\a\_work\1\s\payload\base\clrjit_unix_x64_x64.dll

Download: https://clrjit2.blob.core.windows.net/jitrollingbuild/builds/4198a8af7caa3f9f28db68a90ef4b318f7edb50c/windows/x64/Checked/clrjit_win_x64_x64.dll -> D:\a\_work\1\s\payload\base\clrjit_win_x64_x64.dll

...

ConnectionResetError: [WinError 10054] An existing connection was forcibly closed by the remote host
```
(I've noticed this started happening quite often in the last few days)

**Comment by jakobbotsch** [comment]
Any reason not to spill for the other patterns as well? Then we could remove the the special handling for byref-like types before this function is called.

**Comment by SingleAccretion** [comment]
> Any reason not to spill for the other patterns as well?

Nothing that I would think of I suppose, I'll make the changes.

**Comment by SingleAccretion** [comment]
I have scaled back the change to the narrow correctness fix, as tying together the many ways of how unused values can be transformed while avoiding regressions proved a little fruitless.

<details>
<summary>Details</summary>

There are 4 ways an unused tree can be transformed:

1) Struct, into an address.
2) Struct load, into a nullcheck.
3) Primitive, left alone.
4) Primitive, wrapped in a `COMMA`.

It so happens that for different trees the different strategies produce different results CQ-wise, due to the differences on how friendly a given shape is to CSE and morph's dead code removal. This area is full of quirky behavior, so I decided to leave it be for now.

</details>

**Comment by SingleAccretion** [comment]
The installer timeout looks unlikely to be related; will request a re-run.

@dotnet/jit-contrib

**Comment by jakobbotsch** [comment]
That one is #69114 so not related.

**Review by jakobbotsch — APPROVED** [review]

**Pull request merged**